source/agora/network/NodeLocator.d

```diff
@@ -0,0 +1,268 @@
+    {
+        if (this.mmdb != MMDB_s.init)
+        {
+            fprintf(stderr, "NodeLocator.shutdown() was not called manually");
+            this.stop();
+        }
+    }
+
+    /***************************************************************************
+
+        Extracts certain properties of the geographical location of the node.
+        The properties can be continent, country, ...
+        The exact properties that can be extracted and the query path that
+        identifies the properties are implementation dependent.
+
+        An example query path can be:
+            continent->names->en
+            location->latitude
+
+        Params:
+            address = the network address of the node
+            paths = query parts to identify which properties needs to be extracted
+
+        Returns:
+            an array of extracted properties, if a particular property
+            cannot be extracted, then `MissingValue` is returned
+
+    ***************************************************************************/
+
+    public string[] extractValues (Address address, string[] paths)
+    {
+        import std.algorithm : map;
+        import std.array : array, split;
+        import std.range : repeat;
+        import std.string : toStringz;
+
+        string[] extractedValues;
+        extractedValues.length = paths.length;
+
+        // Look up the value in the database
+        int gai_error;
+        int mmdb_error;
+        MMDB_lookup_result_s result = MMDB_lookup_string(&this.mmdb, address.ptr, &gai_error, &mmdb_error);
+        if (0 != gai_error)
+        {
+            log.warn("Error from getaddrinfo for {}", address);
+            return MissingValue.repeat(paths.length).array();
+        }
+        if (MMDB_SUCCESS != mmdb_error)
+        {
+            log.warn("Got an error from libmaxminddb: {}", MMDB_strerror(mmdb_error));
+            return MissingValue.repeat(paths.length).array();
+        }
+
+        // Extract the data
+        foreach (ind, path; paths)
+        {
+            auto path_splitted = path.split(PathSeparator).map!(path_part => path_part.toStringz()).array();
+            path_splitted ~= null;
+            MMDB_entry_data_s entry_data;
+            int status = MMDB_aget_value(&result.entry, &entry_data, path_splitted.ptr);
+            extractedValues[ind] = (MMDB_SUCCESS == status && entry_data.has_data)
+                ? getStringFromEntry(entry_data)
+                : MissingValue;
+        }
+        return extractedValues;
+    }
+
+    /***************************************************************************
+
+        Convert data stored in MMDB_entry_data_s structure into a string
+
+        Params:
+            entry_data = data to convert
+
+        Returns:
+            data converted to string
+
+    ***************************************************************************/
+
+    private string getStringFromEntry (in MMDB_entry_data_s entry_data) const
```

Comment:
This doesn't use anything from the context and could be `static` or a free function right ?